File: FizzBuzz/java/src/test/java/FizzBuzzTests.java
import org.junit.Test;
import org.junit.After;
import org.junit.Before;
import static org.junit.Assert.assertEquals;

public class FizzBuzzTests {

  private final FizzBuzz fizzBuzz = FizzBuzz.getInstance();

  @Before
  public void initialize() {
      FiltersProvider.add(Fizz.getInstance());
      FiltersProvider.add(Buzz.getInstance());
  }

  @After
  public void cleanUp() {
      FiltersProvider.removeAll();
      PrintersProvider.removeAll();
  }

  @Test
  public void stringFor_whenInputNotDivisibleBy3_or5_thenReturnInputAsString()
      throws Exception {
    final int input = 1;
    final String expected = "1";

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test
  public void stringFor_whenInputDivisibleBy3_thenReturnFizz()
      throws Exception {
    final int input = 3;
    final String expected = "Fizz";

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test
  public void stringFor_whenInputDivisibleBy5_thenReturnBuzz()
      throws Exception {
    final int input = 5;
    final String expected = "Buzz";

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test
  public void stringFor_whenInputDivisibleBy3_and5_thenReturnFizzBuzz()
      throws Exception {
    final int input = 15;
    final String expected = "FizzBuzz";

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test(expected = Exception.class)
  public void stringFor_whenInputLessThan1_thenThrowException()
      throws Exception {
    final int input = 0;

    fizzBuzz.stringFor(input);
  }

  @Test
  public void stringFor_givenListOfInteger_andDefaultPrinter_thenReturnResultAsWhiteSpaceSeparatedString()
      throws Exception {
    final int [] input = new int [] {1,3,5,7,15};
    final String expected = "1 Fizz Buzz 7 FizzBuzz";
    PrintersProvider.add(DefaultPrinter.getInstance());

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test
  public void stringFor_givenListOfInteger_andCommaSeparatedPrinter_thenReturnResultAsCommaSeparatedString()
      throws Exception {
    final int [] input = new int [] {1,3,5,7,15};
    final String expected = "1, Fizz, Buzz, 7, FizzBuzz";
    PrintersProvider.add(CommaSeparatedPrinter.getInstance());

    final String actual = fizzBuzz.stringFor(input);

    assertEquals(expected, actual);
  }

  @Test
  public void stringFor_whenInputIsDivisibleBy2_thenReturnFoo()
      throws Exception {
      final int input = 2;
      final String expected = "Foo";
      final Filter filter = Foo.getInstance();
      FiltersProvider.add(filter);

      final String actual = fizzBuzz.stringFor(input);

      assertEquals(expected, actual);
  }

  @Test
  public void newFilterFoo_ExcatlyOnlyOneInstanceInJVM_canBeInstantiated()
  {
      final Filter filterFoo1 = Foo.getInstance();
      final Filter filterFoo2 = Foo.getInstance();

      assertEquals(filterFoo1, filterFoo2);
  }

  @Test
  public void newFilterFizz_ExcatlyOnlyOneInstanceInJVM_canBeInstantiated()
  {
      final Filter filterFizz1 = Fizz.getInstance();
      final Filter filterFizz2 = Fizz.getInstance();

      assertEquals(filterFizz1, filterFizz2);
  }

  @Test
  public void newFilterBuzz_ExcatlyOnlyOneInstanceInJVM_canBeInstantiated()
  {
      final Filter filterBuzz1 = Buzz.getInstance();
      final Filter filterBuzz2 = Buzz.getInstance();

      assertEquals(filterBuzz1, filterBuzz2);
  }
}

File: FizzBuzz/kotlin/src/test/kotlin/FizzBuzzTests.kt
import org.junit.Test
import org.junit.Before
import org.junit.After
import kotlin.test.assertEquals

class FizzBuzzTest {

	@Before
	fun setup()
	{
		FiltersProvider.add(Fizz)
		FiltersProvider.add(Buzz)
		FizzBuzz.addPrinter(DefaultPrinter)
	}

	@After
	fun cleanUp()
	{
		FiltersProvider.removeAll()
		FizzBuzz.removeAllPrinters()
	}

	@Test(expected = IntegerLessThanOneException::class)
	fun `'stringFor' when input is less than 1, then throw exception`()
	{
		val input = 0

		FizzBuzz.stringFor(input)
	}

	@Test
	fun `'stringFor' when input is divisible by 3, then return 'Fizz'`()
	{
		val input = 3
		val expected = "Fizz"

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' when input is divisible by 5, then return 'Buzz'`()
	{
		val input = 5
		val expected = "Buzz"

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' when input is divisible by 3 and 5, then return 'FizzBuzz'`()
	{
		val input = 15
		val expected = "FizzBuzz"

		val actual =  FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' when input is divisible by 2, then return 'Foo'`()
	{
		val input = 2
		val expected = "Foo"
		FiltersProvider.add(Foo)

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' given list of integer and Default Printer, return result as whitespace-separated string`()
	{
		val input = listOf(1,3,5,7,15)
		val expected = "1 Fizz Buzz 7 FizzBuzz"

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' when input is not divisible by 3 or 5, then return input as String`()
	{
		val input = 1
		val expected = "1"

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

	@Test
	fun `'stringFor' given list of integer and comma separated printer, then return result as a comma-separated string for each integer's divisibility`() {
		val input = mutableListOf(1,3,5,7,15)
		val expected = "1, Fizz, Buzz, 7, FizzBuzz"
		FizzBuzz.addPrinter(CommaSeparatedPrinter)

		val actual = FizzBuzz.stringFor(input)

		assertEquals(expected, actual)
	}

}
